sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -104,3 +104,32 @@ object BindReferences extends Logging {
     }.asInstanceOf[A] // Kind of a hack, but safe.  TODO: Tighten return type when possible.
   }
 }
+
+/**
+ * A column vector reference points to a specific column for ColumnVector.
+ * columnVar is a variable that keeps ColumnVector, and ordinal is row index in ColumnVector
+ */
+case class ColumnVectorReference(
+  columnVar: String, ordinal: String, dataType: DataType, nullable: Boolean)
+  extends LeafExpression {
+
+  override def toString: String = s"columnVector[$columnVar, $ordinal, ${dataType.simpleString}]"
+
+  override def eval(input: InternalRow): Any = null
+
+  override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
+    val javaType = ctx.javaType(dataType)
+    val value = ctx.getValue(columnVar, dataType, ordinal)
+    if (nullable) {
+      s"""
+        boolean ${ev.isNull} = ${columnVar}.isNullAt($ordinal);
+        $javaType ${ev.value} = ${ev.isNull} ? ${ctx.defaultValue(dataType)} : ($value);
```

Comment:
Let me confirm you suggestion. Is your suggestion to move the definition of ```ColumnVectorReference``` to ```ExistingRDD.scala```?  Or, is this to move ```genCode``` into ```DataSourceScan``` by dropping ```ColumnVectorReference``` and copy ```Expression.gen()``` into ```DataSourceScan```?